File: spare_parts_app_new/lib/models/purchase.dart
import 'purchase_item.dart';

class Purchase {
  final int? id;
  final String supplierName;
  final String? supplierMobile;
  final String invoiceNumber;
  final DateTime purchaseDate;
  final List<PurchaseItem> items;
  final double? discount;
  final double totalAmount;
  final String? notes;
  final String? billImageUrl;
  final String? billPdfUrl;
  final double? dailyAmount;
  final double? remainingAmount;
  final int? createdById;
  final String? createdByName;
  final DateTime? createdAt;

  Purchase({
    this.id,
    required this.supplierName,
    this.supplierMobile,
    required this.invoiceNumber,
    required this.purchaseDate,
    required this.items,
    this.discount,
    required this.totalAmount,
    this.notes,
    this.billImageUrl,
    this.billPdfUrl,
    this.dailyAmount,
    this.remainingAmount,
    this.createdById,
    this.createdByName,
    this.createdAt,
  });

  factory Purchase.fromJson(Map<String, dynamic> json) {
    return Purchase(
      id: json['id'],
      supplierName: json['supplierName'] ?? '',
      supplierMobile: json['supplierMobile'],
      invoiceNumber: json['invoiceNumber'] ?? '',
      purchaseDate: json['purchaseDate'] != null 
          ? DateTime.parse(json['purchaseDate']) 
          : DateTime.now(),
      items: (json['items'] as List? ?? [])
          .map((i) => PurchaseItem.fromJson(i))
          .toList(),
      discount: (json['discount'] as num?)?.toDouble(),
      totalAmount: (json['totalAmount'] as num? ?? 0).toDouble(),
      notes: json['notes'],
      billImageUrl: json['billImageUrl'],
      billPdfUrl: json['billPdfUrl'],
      dailyAmount: (json['dailyAmount'] as num?)?.toDouble(),
      remainingAmount: (json['remainingAmount'] as num?)?.toDouble(),
      createdById: json['createdById'],
      createdByName: json['createdByName'],
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'supplierName': supplierName,
      'supplierMobile': supplierMobile,
      'invoiceNumber': invoiceNumber,
      'purchaseDate': purchaseDate.toIso8601String().split('T')[0],
      'items': items.map((i) => i.toJson()).toList(),
      'discount': discount,
      'totalAmount': totalAmount,
      'notes': notes,
      'billImageUrl': billImageUrl,
      'billPdfUrl': billPdfUrl,
      'dailyAmount': dailyAmount,
      'remainingAmount': remainingAmount,
    };
  }
}

File: backend/src/main/java/com/spareparts/inventory/entity/Purchase.java
package com.spareparts.inventory.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "purchases")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Purchase {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(name = "supplier_name")
    private String supplierName;

    @Column(name = "supplier_mobile")
    private String supplierMobile;

    @NotBlank
    @Column(name = "invoice_number")
    private String invoiceNumber;

    @NotNull
    @Column(name = "purchase_date")
    private LocalDate purchaseDate;

    @OneToMany(mappedBy = "purchase", cascade = CascadeType.ALL, orphanRemoval = true)
    private java.util.List<PurchaseItem> items = new java.util.ArrayList<>();

    @Column(name = "discount", precision = 10, scale = 2)
    private BigDecimal discount = BigDecimal.ZERO;

    @NotNull
    @Column(name = "total_amount", precision = 10, scale = 2)
    private BigDecimal totalAmount;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "bill_image_url")
    private String billImageUrl;

    @Column(name = "bill_pdf_url")
    private String billPdfUrl;

    @Column(name = "cost_price", precision = 10, scale = 2)
    private BigDecimal costPrice = BigDecimal.ZERO;

    @Column(name = "daily_amount", precision = 10, scale = 2)
    private BigDecimal dailyAmount;

    @Column(name = "remaining_amount", precision = 10, scale = 2)
    private BigDecimal remainingAmount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id")
    private User createdBy;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
